### PR TITLE
PS-1125: Move all CONSTRAINT/FOREIGN KEY declarations to ALTER TABLE

### PR DESCRIPTION
**Pull request by inikep**
This commit also fixes PS-3863.
It removes fixes for PS-605 which are not needed now.

### CODE REVIEW

File: client/mysqldump.c [hunk 1]
```diff
@@ -2723,18 +2649,25 @@ static void skip_secondary_keys(char *create_str, my_bool has_pk)
       contains_autoinc_column(autoinc_column, ptr, type) : FALSE;
 
     /* Is it a secondary index definition? */
-    if (c == '\n' &&
+    if (c == '\n' && !has_autoinc &&
```
Comment (percona-ysorokin) [inline client/mysqldump.c:2651]: Why moving `has_autoinc` before checking `type`?

File: client/mysqldump.c [hunk 2]
```diff
@@ -3782,27 +3715,47 @@ static void dump_skipped_keys(const char *table)
...
+      my_free(def);
+      my_free(node);
+    }
+  }
+
+  if (alter_constraints_list)
   {
-    LIST *node= skipped_keys_list;
-    char *def= node->data;
+	alter_constraints_list= list_reverse(alter_constraints_list);
```
Comment (percona-ysorokin) [inline client/mysqldump.c]: Indentation

File: client/mysqldump.c [hunk 3]
```diff
@@ -3782,27 +3715,47 @@ static void dump_skipped_keys(const char *table)
...
               table);
 
-  skipped_keys_list= list_reverse(skipped_keys_list);
-  fprintf(md_result_file, "ALTER TABLE %s ", table);
-  for (keys= list_length(skipped_keys_list); keys > 0; keys--)
+  if (skipped_keys_list)
+  {
+    skipped_keys_list= list_reverse(skipped_keys_list);
+    fprintf(md_result_file, "ALTER TABLE %s ", table);
+    for (keys= list_length(skipped_keys_list); keys > 0; keys--)
```
Comment (percona-ysorokin) [inline client/mysqldump.c]: Why do we generate separate `ALTER TABLE` statement for each constraint? Can we combine them into a single statement? E. g.
```
ALTER TABLE t1
  ADD CONSTRAINT t1_ibfk_1 FOREIGN KEY (a) REFERENCES t1 (id),
  ADD CONSTRAINT t1_ibfk_2 FOREIGN KEY (b) REFERENCES t1 (id);
```

Comment (percona-ysorokin) [reply client/mysqldump.c]: Moreover, why separate `CONSTRAINT`s from `KEY`s / `UNIQUE_KEY`s / `PRIMARY KEY`s? I believe MySQL grammar allows mixing them:
```
ALTER TABLE t1
  ADD KEY `a_index` (`a`),
  ADD CONSTRAINT `t1_const` FOREIGN KEY (`a`) REFERENCES `t1` (`id`);
```

File: mysql-test/t/percona_mysqldump_innodb_optimize_keys.test [hunk 4]
```diff
@@ -298,5 +300,97 @@ PARTITIONS 2*/;
...
+
+--exec $MYSQL test < $file
+
+--remove_file $file
+
+DROP TABLE t1;
+
+#############################################################################
+# Bug PS-3863: mysqldump --innodb-optimize-keys may produce invalid SQL with
+#              implicitly named FK constraints
```
Comment (percona-ysorokin) [inline mysql-test/t/percona_mysqldump_innodb_optimize_keys.test]: `explicitly`?

**Review by percona-ysorokin — CHANGES_REQUESTED** [review]

**Comment by percona-ysorokin** [comment]
Jenkins build link
https://jenkins.percona.com/job/percona-server-5.6-param/2086/

File: client/mysqldump.c [hunk 5]
```diff
@@ -204,6 +205,7 @@ TYPELIB compatible_mode_typelib= {array_elements(compatible_mode_names) - 1,
 HASH ignore_table;
 
 LIST *skipped_keys_list;
+LIST *alter_constraints_list;
```
Comment (laurynas-biveinis) [inline client/mysqldump.c]: static

File: mysql-test/t/percona_mysqldump_innodb_optimize_keys.test [hunk 6]
```diff
@@ -246,7 +246,7 @@ DROP TABLE t1;
 
 #############################################################################
 # Bug #1081016: mysqldump --innodb-optimize-keys may produce invalid SQL with
-#               explicitly named FK constraints
+#     PS-605    explicitly named FK constraints
```
Comment (laurynas-biveinis) [inline mysql-test/t/percona_mysqldump_innodb_optimize_keys.test]: Why inserting a new bug number in the middle of another line-wrapped bug title?

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]

File: client/mysqldump.c (hunk 1 above)
Comment (inikep) [reply client/mysqldump.c:2651]: I think it makes code more readable.

**Review by percona-ysorokin — APPROVED** [review]
LGTM

**Comment by inikep** [comment]
https://jenkins.percona.com/job/percona-server-5.5-param/1634/

**Review by laurynas-biveinis — APPROVED** [review]

**Pull request merged**